### DISCUSSION TITLE
How to confirm that two subscriptions are from the same user?

### DISCUSSION THREAD

**Discussion by caoyongfeng0214** [hyochan/react-native-iap#2903] [Q&A]
On iOS, each subscription has the same originalTransactionIdentifierIOS, which can be used to confirm whether it comes from the same user.

However, there is no similar identifier in the purchaseHistory and currentPurchase objects in Android.

Can purchaseToken be used to make this judgment?

If a user subscribes to "Subscription A" (paid monthly), and it automatically renews after one month, is the purchaseToken used during renewal the same as the previous purchaseToken?

If a user subscribes to "Subscription A" (paid monthly), and one month later he wants to change to "Subscription B" (paid annually), is the purchaseToken used during subscription to "Subscription B" the same as the previous purchaseToken?

Thanks.

**Comment by hyochan** [comment]
On iOS, `originalTransactionIdentifierIOS` stays the same across all renewals, making it easy to identify the same user.

On Android, `purchaseToken` behaves differently.

## purchaseToken Behavior

### Scenario 1: Auto-renewal

**purchaseToken stays the SAME.**

When a subscription renews automatically, the same `purchaseToken` is retained. No new token is generated.

### Scenario 2: Upgrade/Downgrade (e.g., Monthly → Annual)

**purchaseToken CHANGES.**

A new `purchaseToken` is generated. The old token is invalidated.

To link them, use `linkedPurchaseToken` from the Google Play Developer API (server-side).

### Scenario 3: Resubscribe after cancellation (before expiry)

**purchaseToken CHANGES.**

New token generated with `linkedPurchaseToken` pointing to the old token.

### Scenario 4: Resubscribe after expiration

**purchaseToken CHANGES.**

New token generated, but `linkedPurchaseToken` is NOT set (original subscription fully expired).

## Solution: Use linkedPurchaseToken (Server-side)

Google Play Developer API returns `linkedPurchaseToken` which points to the previous token when:
- User upgrades/downgrades
- User resubscribes before expiry

```
Token A (original) ← linkedPurchaseToken ← Token B (upgrade)
```

**Important**: `linkedPurchaseToken` is only available via the server-side API, not in the client-side purchase object.

## Easy Solution: Use IAPKit with verifyPurchaseWithProvider

If you don't want to build your own server, you can use [IAPKit](https://iapkit.com) with `verifyPurchaseWithProvider()`. IAPKit handles `linkedPurchaseToken` tracking automatically and provides a unified API for both iOS and Android.

```typescript
import { verifyPurchaseWithProvider } from 'react-native-iap';

const result = await verifyPurchaseWithProvider({
  provider: 'iapkit',
  iapkit: {
    apiKey: 'your-iapkit-api-key',
    google: { purchaseToken: purchase.purchaseToken! },
  },
});

// IAPKit returns linkedPurchaseToken info when available
if (result.iapkit?.isValid) {
  console.log('Purchase verified:', result.iapkit.state);
}
```

## Manual Approach

1. **Server-side validation** - Always verify purchases on your server
2. **Store user mapping** - Map `purchaseToken` to your user ID on first purchase
3. **Follow linkedPurchaseToken chain** - When you see a new token with `linkedPurchaseToken`, link it to the same user
4. **Use obfuscatedAccountIdAndroid** - Pass your user ID when making purchases:

```typescript
await requestPurchase({
  request: {
    google: {
      skus: ['subscription_id'],
      obfuscatedAccountIdAndroid: 'your_user_id_hash',
    },
  },
  type: 'subs',
});
```

## References

- [Google: linkedPurchaseToken Implementation](https://medium.com/androiddevelopers/implementing-linkedpurchasetoken-correctly-to-prevent-duplicate-subscriptions-82dfbf7167da)
- [Google: Subscription Lifecycle](https://developer.android.com/google/play/billing/lifecycle/subscriptions)
- [Google: About Subscriptions](https://developer.android.com/google/play/billing/subscriptions)
